### PR TITLE
Windows CI implementation

### PR DESCRIPTION
**Pull request by NexeyaSGara**
This PR adds:
- windows CI with appveyor
- CMakelist based project
- VC15 compatibility

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,219 @@
+cmake_minimum_required(VERSION 3.0.2)
+set(CMP0048 NEW)
+
+project(_tango LANGUAGES CXX
+        VERSION 1.0)
+        
+#need to define the version of the library
+set(MAJOR_VERSION "9")
+set(MINOR_VERSION "2")
+set(PATCH_VERSION "2")
```
Comment (vxgmichel) [inline CMakeLists.txt]: The patch version needs to be updated :)

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,324 @@
+version: 1.0.{build}
+branches:
+  only:
+    - NexeyaSGara-windows-ci
```
Comment (vxgmichel) [inline appveyor.yml:4]: This needs to be updated.

File: appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+  - del "C:\Program Files (x86)\MSBuild\4.0\Microsoft.Common.Targets\ImportAfter\Xamarin.Common.targets"
+  - del "C:\Program Files (x86)\MSBuild\14.0\Microsoft.Common.targets\ImportAfter\Xamarin.Common.targets"
+  - del "C:\Program Files (x86)\MSBuild\12.0\Microsoft.Common.targets\ImportAfter\Xamarin.Common.targets"
+  #RDP from start
+  #- ps: iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/appveyor/ci/master/scripts/enable-rdp.ps1'))
+  # Tango
+  - cmd: cd "C:\projects\"
+  - cmd: md libtango
+  - cmd: cd "C:\projects\" 
+  - if %ARCH_PY%==win32-msvc9-py27 appveyor DownloadFile https://ci.appveyor.com/api/buildjobs/fibcw0rnx1ahit1d/artifacts/libtango_9.2.6_v90_x86.zip
```
Comment (vxgmichel) [inline appveyor.yml:161]: Can we do something about the hard-coded artifacts?

File: ext/api_util.cpp [hunk 4]
```diff
@@ -15,6 +15,18 @@
...
+namespace boost
+{
+	template <>
+	Tango::ApiUtil const volatile * get_pointer<class Tango::ApiUtil const volatile >(
+		class Tango::ApiUtil const volatile *c)
+	{
+		return c;
+	}
+}
+#endif
```
Comment (vxgmichel) [inline ext/api_util.cpp]: Maybe this code block can be refactored to be included only once?

File: tango/release.py [hunk 5]
```diff
@@ -49,7 +49,7 @@ class Release:
     version = '.'.join(map(str, version_info[:3]))
     release = ''.join(map(str, version_info[3:]))
     separator = '.' if 'dev' in release or 'post' in release else ''
-    version_long = version + separator + release
+    version_long = "9.2.3"
```
Comment (vxgmichel) [inline tango/release.py]: Has this change been committed by mistake?

**Review by vxgmichel — COMMENTED** [review]
This is not a complete review, just a few comments.

File: appveyor.yml (hunk 3 above)
Comment (NexeyaSGara) [reply appveyor.yml:161]: Now, we can use the GitHub releases, but I don't know if we can have a "last release" think

File: tango/release.py (hunk 5 above)
Comment (NexeyaSGara) [reply tango/release.py]: I think I had a problem somewhere and tested it like that and it stayed,

File: ext/api_util.cpp (hunk 4 above)
Comment (NexeyaSGara) [reply ext/api_util.cpp]: Surely, but I don't know where could be the best place to add it.

File: ext/api_util.cpp [hunk 6]
```diff
@@ -15,6 +15,18 @@
 
 using namespace boost::python;
 
+#if _MSC_VER > 1800
+namespace boost
+{
+	template <>
+	Tango::ApiUtil const volatile * get_pointer<class Tango::ApiUtil const volatile >(
```
Comment (sdebionne) [inline ext/api_util.cpp]: At first, I thought this was invalid C++. It's very unusual to use [elaborated type specifiers](https://en.cppreference.com/w/cpp/language/elaborated_type_specifier). And I think you don't need to [specify the specialization explicitely](https://en.cppreference.com/w/cpp/language/template_specialization#Explicit_specializations_of_function_templates).
If you want a full specialization of `boost::get_pointer<>()` for ` Tango::ApiUtil const volatile *`, then use:
```c++
Tango::ApiUtil const volatile * get_pointer<>(Tango::ApiUtil const volatile *c) { return c; }
```
But how is this different from:
```c++
// Extract from boost/get_pointer.hpp
template<class T> T * get_pointer(T * p)
{
    return p;
}
```
And why only ` _MSC_VER > 1800` requires it? MSVC bug?

File: CMakeLists.txt (hunk 1 above)
Comment (sdebionne) [reply CMakeLists.txt]: There are ways to tell CMake to get the project version from git tags, if you are interested, I can provide a  PR.

File: ext/api_util.cpp (hunk 6 above)
Comment (NexeyaSGara) [reply ext/api_util.cpp]: yes MSVC14 bug

File: tango/__init__.py [hunk 7]
```diff
@@ -121,12 +121,18 @@ def __prepare_nt():
...
     else:
-        # Tango C++ could not be found on the system...
-        # ... use PyTango's private Tango C++ library
-        tango_dll_path = os.path.dirname(os.path.abspath(__file__))
-        tango_dll_path = os.path.join(tango_dll_path, "_tango_dll_")
-        if os.path.exists(tango_dll_path):
+        tango_dll_path = os.path.join(tango_root, "bin")
+        tango_dll_path = tango_dll_path.lower()
+        if os.path.exists(tango_dll_path) and \
+           tango_dll_path not in PATH.lower():
```
Comment (ajoubertza) [inline tango/__init__.py:127]: I think the check that it is in PATH variable already should be a separate condition inside this `if` block.  If the path exists, but isn't in the PATH variable already, then we won't add it - we just fall back to the the "Tango C++ could not be found".  That seems incorrect.  Rather:
```
        if os.path.exists(tango_dll_path):
            if tango_dll_path not in PATH.lower():
                os.environ['PATH'] += ";" + tango_dll_path
        else:
```

File: CMakeLists.txt (hunk 1 above)
Comment (ajoubertza) [reply CMakeLists.txt]: I would like this to be derived from the git tag.

File: appveyor.yml (hunk 2 above)
Comment (ajoubertza) [reply appveyor.yml:4]: Agree.  Obviously only useful to change once the PR is about to be merged.  Would it be `develop`?  Ideally we want releases to be built from tags, in which case the branch doesn't matter.   I'm not sure how that would be done in appveyor.

Having it run from `develop` would also be useful, just to see if things are still working or breaking.

File: ext/api_util.cpp (hunk 6 above)
Comment (ajoubertza) [reply ext/api_util.cpp]: Does the bug also exist in newer versions?  Maybe we want this check to also have an upper limit.

**Review by ajoubertza — COMMENTED** [review]
Wow, this has been open for more than year!  Hope we can wrap it up this week.

File: tango/release.py [hunk 8]
```diff
@@ -45,11 +45,11 @@ class Release:
         - license: (str) the license
     """
     name = 'pytango'
-    version_info = (9, 2, 3, 'dev', 0)
+    version_info = (9, 2, 3, 'alpha', 0)  # alpha for fixing windows build
     version = '.'.join(map(str, version_info[:3]))
     release = ''.join(map(str, version_info[3:]))
     separator = '.' if 'dev' in release or 'post' in release else ''
```
Comment (ajoubertza) [inline tango/release.py:51]: This "dev" will need to change to "alpha".
Also, merge the develop branch into yours to resolve the merge conflict.
After that, find and replace all the references to "dev" - for example, in doc/how-to-contribute.rst

**Review by ajoubertza — CHANGES_REQUESTED** [review]
This is looking better with all the special cases in one file.

File: tango/release.py (hunk 8 above)
Comment (ajoubertza) [reply tango/release.py:51]: @NexeyaSGara are you still working on this? We'd like to include it in the 9.3.0 release.

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        VERSION 1.0)
+        
+#need to define the version of the library
+set(MAJOR_VERSION "9")
+set(MINOR_VERSION "2")
+set(PATCH_VERSION "6")
+
+#convenient versions
+set(LIBRARY_VERSION "${MAJOR_VERSION}.${MINOR_VERSION}.${PATCH_VERSION}")
+if(NOT WIN32)
```
Comment (ajoubertza) [inline CMakeLists.txt:14]: This file is being added specifically for Windows builds, so is it really useful for other platforms?  If not useful, I suggest we just abort at this stage:
```
if(NOT WIN32)
message(FATAL_ERROR "This file is only targeting Windows builds - aborting!")
endif(NOT WIN32)
```

**Comment by tiagocoutinho** [comment]
@ajoubertza,
I have been doing the final touches to release 9.3.0.
Is it ok to leave this PR for the 9.3.1? I volunteer to do a 9.3.1 as soon as this is ready.

**Comment by ajoubertza** [comment]
@tiagocoutinho Yes, this will have to wait for 9.3.1, or later.  I thought the remaining changes would be trivial, but they are not.

**Comment by tiagocoutinho** [comment]
Ok, many thanks!

**Comment by ajoubertza** [comment]
Closing this as it has be superseded by PR #277.

**Pull request closed without merging**